File: iosApp/iosApp/viewmodel/PhoneNumberViewModel.swift
import Foundation
import SwiftUI
import Shared

@MainActor
class PhoneNumberViewModel: ObservableObject {
    @Published var phoneNumber: String = ""
    @Published var countryCode: String = "+91"
    @Published var isLoading: Bool = false
    @Published var isListening: Bool = false
    @Published var isProcessing: Bool = false
    @Published var error: String?
    
    private let authRepository: AuthRepository
    private let sendOtpUseCase: SendOtpUseCase
    private let speechRecognizer: IOSSpeechRecognizerWrapper
    
    init() {
        self.authRepository = IOSAuthRepository()
        self.sendOtpUseCase = SendOtpUseCase(authRepository: authRepository)
        self.speechRecognizer = IOSSpeechRecognizerWrapper()
    }
    
    func startSpeechRecognition() async {
        guard !isListening else { return }
        
        isListening = true
        error = nil
        
        // Haptic feedback
        let impactFeedback = UIImpactFeedbackGenerator(style: .medium)
        impactFeedback.impactOccurred()
        
        do {
            // Request permission if needed
            if !speechRecognizer.hasPermission() {
                let granted = try await speechRecognizer.requestPermission()
                if !granted {
                    isListening = false
                    error = "Speech recognition permission is required"
                    return
                }
            }
            
            // Start listening with continuous updates
            Task {
                do {
                    try await speechRecognizer.startListening { [weak self] recognizedText in
                        guard let self = self else { return }
                        // Extract digits from recognized text and update continuously
                        let digits = recognizedText.filter { $0.isNumber }
                        if !digits.isEmpty {
                            self.phoneNumber = digits
                        }
                    }
                } catch {
                    if !error.localizedDescription.contains("cancelled") {
                        self.error = (error as? Error)?.localizedDescription ?? "Speech recognition error"
                    }
                    self.isListening = false
                }
            }
        } catch {
            self.error = (error as? Error)?.localizedDescription ?? "Speech recognition error"
            isListening = false
        }
    }
    
    func stopSpeechRecognition() async {
        guard isListening else { return }
        
        // Haptic feedback
        let impactFeedback = UIImpactFeedbackGenerator(style: .light)
        impactFeedback.impactOccurred()
        
        // Set processing state
        isListening = false
        isProcessing = true
        
        let initialText = speechRecognizer.stopListening()
        // Extract digits from initial text (will be updated via callback if final result comes)
        let digits = initialText.filter { $0.isNumber }
        if !digits.isEmpty {
            phoneNumber = digits
        }
        
        // Wait a bit for final results to come in (up to 3 seconds)
        try? await Task.sleep(nanoseconds: 3_500_000_000) // 3.5 seconds
        
        // Get final accumulated text after waiting
        let finalText = speechRecognizer.getAccumulatedText()
        let finalDigits = finalText.filter { $0.isNumber }
        if !finalDigits.isEmpty {
            phoneNumber = finalDigits
        }
        
        isProcessing = false
    }
    
    func sendOtp(onSuccess: @escaping (String) -> Void, onError: @escaping (String) -> Void) async {
        guard phoneNumber.count >= 10 else {
            error = "Please enter a valid phone number"
            return
        }
        
        let fullPhoneNumber = "\(countryCode)\(phoneNumber)"
        isLoading = true
        error = nil
        
        do {
            print("[PhoneNumberViewModel] Calling sendOtpUseCase for: \(fullPhoneNumber)")
            try await sendOtpUseCase.invoke(phoneNumber: fullPhoneNumber)
            print("✅ [PhoneNumberViewModel] OTP sent successfully, calling onSuccess")
            isLoading = false
            onSuccess(fullPhoneNumber)
            print("✅ [PhoneNumberViewModel] onSuccess callback completed")
        } catch {
            print("❌ [PhoneNumberViewModel] Error sending OTP: \(error)")
            isLoading = false
            let errorMessage = error.localizedDescription
            self.error = errorMessage
            onError(errorMessage)
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/PhoneNumberViewModel.kt
package com.kissangram.viewmodel

import android.app.Activity
import android.app.Application
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.AndroidSpeechRepository
import com.kissangram.usecase.SendOtpUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class PhoneNumberViewModel(
    application: Application,
    private val activity: Activity? = null
) : AndroidViewModel(application) {
    
    private val preferencesRepository = AndroidPreferencesRepository(application)
    private val authRepository = AndroidAuthRepository(
        context = application,
        activity = activity,
        preferencesRepository = preferencesRepository
    )
    private val sendOtpUseCase = SendOtpUseCase(authRepository)
    private val speechRepository = AndroidSpeechRepository(application)
    
    private val _uiState = MutableStateFlow(PhoneNumberUiState())
    val uiState: StateFlow<PhoneNumberUiState> = _uiState.asStateFlow()
    
    fun updatePhoneNumber(phoneNumber: String) {
        _uiState.value = _uiState.value.copy(
            phoneNumber = phoneNumber,
            error = null
        )
    }
    
    fun updateCountryCode(countryCode: String) {
        _uiState.value = _uiState.value.copy(countryCode = countryCode)
    }
    
    fun startSpeechRecognition() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isListening = true, error = null)
            
            // Check permission
            val hasPermission = ContextCompat.checkSelfPermission(
                getApplication(),
                android.Manifest.permission.RECORD_AUDIO
            ) == PackageManager.PERMISSION_GRANTED
            
            if (!hasPermission) {
                _uiState.value = _uiState.value.copy(
                    isListening = false,
                    error = "Microphone permission is required for speech recognition"
                )
                return@launch
            }
            
            // Set up callback for continuous updates BEFORE starting to listen
            speechRepository.setOnTextUpdate { recognizedText ->
                // Extract digits from recognized text and update continuously
                // Callback is already on main thread via Handler
                val digits = recognizedText.filter { it.isDigit() }
                if (digits.isNotEmpty()) {
                    _uiState.value = _uiState.value.copy(
                        phoneNumber = digits,
                        error = null
                    )
                }
            }
            
            // Start listening in background - it will continue until stopped
            try {
                speechRepository.startListening()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isListening = false,
                    error = e.message ?: "Speech recognition error"
                )
            }
        }
    }
    
    fun stopSpeechRecognition() {
        android.util.Log.d("PhoneNumberViewModel", "stopSpeechRecognition called")
        viewModelScope.launch {
            // Set processing state
            _uiState.value = _uiState.value.copy(isListening = false, isProcessing = true)
            android.util.Log.d("PhoneNumberViewModel", "Calling speechRepository.stopListening()")
            speechRepository.stopListening()
            // Wait a bit for final results to come in (up to 3 seconds)
            kotlinx.coroutines.delay(3500) // Wait 3.5 seconds to ensure final results are processed
            // Get final accumulated text
            val finalText = speechRepository.getAccumulatedText()
            val digits = finalText.filter { it.isDigit() }
            if (digits.isNotEmpty()) {
                _uiState.value = _uiState.value.copy(phoneNumber = digits)
            }
            _uiState.value = _uiState.value.copy(isProcessing = false)
        }
    }
    
    fun sendOtp(onSuccess: (String) -> Unit, onError: (String) -> Unit) {
        val phone = _uiState.value.phoneNumber
        if (phone.isBlank() || phone.length < 10) {
            _uiState.value = _uiState.value.copy(
                error = "Please enter a valid phone number"
            )
            return
        }
        
        val fullPhoneNumber = "${_uiState.value.countryCode}$phone"
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            
            try {
                sendOtpUseCase(fullPhoneNumber)
                _uiState.value = _uiState.value.copy(isLoading = false)
                onSuccess(fullPhoneNumber)
            } catch (exception: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = exception.message ?: "Failed to send OTP"
                )
                onError(exception.message ?: "Failed to send OTP")
            }
        }
    }
}

data class PhoneNumberUiState(
    val phoneNumber: String = "",
    val countryCode: String = "+91",
    val isLoading: Boolean = false,
    val isListening: Boolean = false,
    val isProcessing: Boolean = false,
    val error: String? = null
)
